src/DnsClient/DnsTcpMessageHandler.cs

```diff
@@ -110,14 +112,23 @@ private async Task<DnsResponseMessage> QueryAsyncInternal(TcpClient client, Netw
 
             do
             {
+                const int LengthSize = 2;
                 int length = 0;
-                try
-                {
-                    length = stream.ReadByte() << 8 | stream.ReadByte();
-                }
-                catch (Exception ex) when (ex is IOException || ex is SocketException)
+                using (var memory = new PooledBytes(LengthSize))
                 {
-                    break;
+                    try
+                    {
+                        var bytesReceived = await stream.ReadAsync(memory.Buffer, 0, LengthSize, cancellationToken)

```

Comment:
hi,
why do you think this is necessary to pool for 2 bytes and read them asynchronously?
That code has some (small) extra overhead just to read 2 bytes ;)

Reply:
The code is also wrong. ReadAsync() won't necessarily return the 2 bytes being requested, so this needs to be in a loop. Something more like this:

```csharp
var lengthBuffer = new byte[2];
int nread = 0, n;

do {
    if ((n = await stream.ReadAsync (lengthBuffer, nread, lengthBuffer.Length - nread, cancellationToken).ConfigureAwait (false)) > 0)
        nread += n;
} while (nread < lengthBuffer.Length);
```

Reply:
That said, what I would recommend doing is buffering the response from the server by reading as much as you can anytime you call out to the kernel to do network I/O. Then, after reading a chunk of data, parse as much as you can, refill the buffer, and continue parsing.

I've got some examples of this approach in the MimeKit (MimeParser) and MailKit (ImapStream, Pop3Stream, and SmtpStream) sources. Look for the ReadAheadAsync() methods and how they are used to refill the input buffer.